File: client/render/sprite.go
package render

const (
	spriteScale = 1.0 / 64.0
)

// Sprite struct
type Sprite struct {
	Atlas     []int
	Width     float32
	HalfWidth float32
	Height    float32
	Left      float32
	Top       float32
	Right     float32
	Bottom    float32
	OX        float32
	OY        float32
}

// SimpleSprite func
func SimpleSprite(left, top, width, height, atlasWidth, atlasHeight float32) [4]float32 {
	return [4]float32{
		left * atlasWidth,
		top * atlasHeight,
		(left + width) * atlasWidth,
		(top + height) * atlasHeight,
	}
}

// SpriteBuild func
func SpriteBuild(atlas []int, atlasWidth, atlasHeight float32) *Sprite {
	s := &Sprite{}
	s.Atlas = atlas

	atlasLeft := float32(atlas[0])
	atlasTop := float32(atlas[1])
	atlasRight := float32(atlas[2])
	atlasBottom := float32(atlas[3])

	s.Left = atlasLeft * atlasWidth
	s.Top = atlasTop * atlasHeight
	s.Right = (atlasLeft + atlasRight) * atlasWidth
	s.Bottom = (atlasTop + atlasBottom) * atlasHeight

	s.Width = atlasRight
	s.Height = atlasBottom
	s.HalfWidth = s.Width * 0.5

	if len(atlas) > 4 {
		s.OX = float32(atlas[4])
		s.OY = float32(atlas[5])
	} else {
		s.OX = 0.0
		s.OY = 0.0
	}

	return s
}

// SpriteBuild3d func
func SpriteBuild3d(atlas []int, atlasWidth, atlasHeight float32) *Sprite {
	s := &Sprite{}
	s.Atlas = atlas

	atlasLeft := float32(atlas[0])
	atlasTop := float32(atlas[1])
	atlasRight := float32(atlas[2])
	atlasBottom := float32(atlas[3])

	s.Left = atlasLeft * atlasWidth
	s.Top = atlasTop * atlasHeight
	s.Right = (atlasLeft + atlasRight) * atlasWidth
	s.Bottom = (atlasTop + atlasBottom) * atlasHeight

	s.Width = atlasRight * spriteScale
	s.Height = atlasBottom * spriteScale
	s.HalfWidth = s.Width * 0.5

	if len(atlas) > 4 {
		s.OX = float32(atlas[4]) * spriteScale
		s.OY = float32(atlas[5]) * spriteScale
	} else {
		s.OX = 0.0
		s.OY = 0.0
	}

	return s
}

File: public/lib/sprite.js
const SpriteScale = 1.0 / 64.0

class Sprite {
    constructor(atlas, width, height, left, top, right, bottom, ox, oy) {
        this.atlas = atlas
        this.width = width
        this.half_width = width * 0.5
        this.height = height
        this.left = left
        this.top = top
        this.right = right
        this.bottom = bottom
        this.ox = ox
        this.oy = oy
    }
    static Simple(left, top, width, height, atlas_width, atlas_height) {
        return [
            left * atlas_width,
            top * atlas_height,
            (left + width) * atlas_width,
            (top + height) * atlas_height
        ]
    }
    static Build(atlas, atlas_width, atlas_height) {
        let left = atlas[0] * atlas_width
        let top = atlas[1] * atlas_height
        let right = (atlas[0] + atlas[2]) * atlas_width
        let bottom = (atlas[1] + atlas[3]) * atlas_height

        let width = atlas[2]
        let height = atlas[3]

        let ox = 0
        let oy = 0
        if (atlas.length > 4) {
            ox = atlas[4]
            oy = atlas[5]
        }

        return new Sprite(atlas, width, height, left, top, right, bottom, ox, oy)
    }
    static Build3(atlas, atlas_width, atlas_height) {
        let left = atlas[0] * atlas_width
        let top = atlas[1] * atlas_height
        let right = (atlas[0] + atlas[2]) * atlas_width
        let bottom = (atlas[1] + atlas[3]) * atlas_height

        let width = atlas[2] * SpriteScale
        let height = atlas[3] * SpriteScale

        let ox = 0
        let oy = 0
        if (atlas.length > 4) {
            ox = atlas[4] * SpriteScale
            oy = atlas[5] * SpriteScale
        }

        return new Sprite(atlas, width, height, left, top, right, bottom, ox, oy)
    }
    static Copy(sprite, ox, oy) {
        return new Sprite(sprite.atlas, sprite.width, sprite.height, sprite.left, sprite.top, sprite.right, sprite.bottom, ox, oy)
    }
}
